Fixes #37015 - add validation error message for host_owner setting

nofaralfasi opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

nofaralfasi commented:
When users attempt to update the `host_owner` setting through Hammer/API, they may not be aware of the accepted value format. 
In the UI, we provide a selection list containing the names of users and usergroups, allowing users to make a choice. The selected value is then saved as `id_and_type`. For example, for the admin user with ID 1, the value would be `1-Users`.

If a username is sent as the value for `host_owner`, an error occurs. In this PR, I propose modifying the error message that occurs in such cases to inform the user about the correct format. 
As discussed with @gauravtalreja1, we are actively seeking the best possible solution here.

ekohl commented on app/models/setting.rb:
```diff
@@ -233,7 +233,10 @@ def validate_host_owner
-    errors.add(:value, _("Host owner is invalid")) if owner.nil?
+    if owner.nil?
+      errors.add(:value, _("Host owner is invalid. " \
+                           "Please ensure the input follows the format 'ID-Users/Usergroups'."))
```
If you're doing this, I think it should be clear that `ID-` is also the prefix for `Usergroups`. Right now that isn't obvious to me. And that `ID` must be replaced with an actual value but `Users` or `Usergroups`. I'd either go with `${id}-[Users|Usergroups]` or `$id-$type where type is Users or Usergroups`.

ekohl reviewed (commented):
We have the `id_and_type` in more places. Is this the only place where it shows up?

I wonder if `OwnerClassifier` should raise an exception instead of silently returning `nil` on invalid input.

ShimShtein commented on app/models/setting.rb:
```diff
@@ -232,8 +232,13 @@ def self.model_name
+
+    begin
+      owner = OwnerClassifier.new(owner_type_and_id).user_or_usergroup
+      raise StandardError, _("Host owner is invalid. Please ensure the input follows the format '${id}-[Users|Usergroups]'.") if owner.nil?
+    rescue StandardError => e
+      errors.add(:value, e.message)
```
Why would you raise an exception and immediately rescue it?
You can add a more detailed message to the `errors` array instead.

ekohl replied:
I suggested that `OwnerClassifier` would raise the exception, not the model.

nofaralfasi commented:
> We have the `id_and_type` in more places. Is this the only place where it shows up?

I didn't find any other places except here.

> I wonder if `OwnerClassifier` should raise an exception instead of silently returning `nil` on invalid input.

While I agree that this is how it should be, this change is causing a lot of errors in our testing.

nofaralfasi commented:
Unless someone has a better idea, I suggest moving the error messages back to the `validate_host_owner` method.

stejskalleos commented:
The CI is failing, not sure if it's due to the changes or just unstable pipeline

nofaralfasi commented:
> The CI is failing, not sure if it's due to the changes or just unstable pipeline

This is due to recent changes. These modifications are raising an error in the `user_or_usergroup` method, which the tests do not anticipate. This is happening because other instances, such as [this](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/hostext/ownership.rb#L47), also invoke the `user_or_usergroup` method.

So my question is, how should we approach this? Should we consider changing the implementation? I don't believe fixing the tests is the solution, given other uses of the `user_or_usergroup` method. We could either raise the exceptions outside the `user_or_usergroup` method but still within the `OwnerClassifier` class, or keep it as it was, adding the error only to the `validate_host_owner` method.

ekohl commented on app/services/owner_classifier.rb:
```diff
@@ -4,7 +4,27 @@ def initialize(id_and_type)
   end
+
+  def valid_input?
+    @id_and_type =~ /^\d+-(Users|Usergroups)$/
```
I know you're refactoring, but while you're at it it's better to use `.match?` if you don't use the captured value. In Ruby 3.2+ you also can't use `nil =~` anymore, or on other non-string values. So this is future proofing it.
```suggestion
    @id_and_type.is_a?(String) && @id_and_type.match?(/^\d+-(Users|Usergroups)$/)
```

ekohl commented on app/services/owner_classifier.rb:
```diff
@@ -4,7 +4,27 @@ def initialize(id_and_type)
   end
 
   def user_or_usergroup
```
I suspect that if you handle `nil` that the test failures will (mostly) disappear. Because AFAIK `Setting[:host_owner]` can be `nil` and that should be handled.
```suggestion
  def user_or_usergroup
    return nil unless @id_and_type

```

nofaralfasi replied:
Thank you for this suggestion!

ekohl commented on app/services/owner_classifier.rb:
I don't want to push you further, but reading all of this, all I can think of is that it should be static. So the API would be `OwnerClassifier.user_or_usergroup(owner_type_and_id)`.

You can consider doing that. Deprecate the current method and make the new method raise exceptions.

ekohl reviewed (commented):
Implicitly these methods can now also raise exceptions:
https://github.com/theforeman/foreman/blob/c10969136f90584dc5ead7e5e3a4a53c938728db/app/models/concerns/hostext/ownership.rb#L37-L49

I wonder if that could cause problems. `Setting[:host_owner]` is validated, but I doubt we have code right now that ensures the ID it uses actually exists. For example, the user or group can probably be deleted, even if it's used in the setting. I don't know if there's a good way to prevent that. Perhaps worth checking with QE how far we want to go here.

ekohl commented on app/services/owner_classifier.rb:
```diff
@@ -1,10 +1,30 @@
+      id_and_type.include?('Users') ? User.find_by_id(id_and_type.to_i) : Usergroup.find_by_id(id_and_type.to_i)
+    end
 
-  def user_or_usergroup
```
You're sure this isn't used anywhere? Normally we add a deprecation and remove it in a future release. Mostly for plugins.

nofaralfasi replied:
I didn't find any other uses in Foreman or Katello. 
I can keep the old method name, but since we are changing it to a static method, we would still need to make adjustments.

ShimShtein replied:
Why not adding a static method in addition to the instance one:

``` ruby
class OwnerClassifier
  def self.classify_owner(owner_id_and_type)
    return self.new(owner_id_and_type).user_or_usergroup
  end

  def user_or_usergroup
    # ...
  end
end
```

This will keep the old way and also add the static shortcut. It will also be much more testable.

nofaralfasi replied:
@ShimShtein, could you check the recent changes and confirm if they match your suggestion?

nofaralfasi committed:
Fixes #37015 - Refactor OwnerClassifier

- Introduce `classify_owner` class method for owner classification.
- Implement private method for input validation.
- Add error handling to raise the necessary exceptions.
- Add deprecation warning to the `user_or_usergroup` method.

ekohl commented:
I think this could be merged, but please also consider this.

Currently you call the old code from the new code. You can also implement the real logic in the new code. This allows you to change the old code to first emit a deprecation and then call the new code

ShimShtein reviewed (dismissed):
LGTM

Maybe we should start using [globalid](https://github.com/rails/globalid) feature of rails instead of reinventing the wheel :thinking:

nofaralfasi commented:
> Maybe we should start using [globalid](https://github.com/rails/globalid) feature of rails instead of reinventing the wheel 🤔

I don't see how using it will help us with the API requests. What values does the user pass when setting the host_owner via API?

stejskalleos reviewed (approved):
:green_apple: **LGTM**

Works as expected, users and usergroup set correctly

stejskalleos commented:
@ekohl could you ACK the changes so it can be released with a follow-up snap?

ekohl commented:
> Currently you call the old code from the new code. You can also implement the real logic in the new code. This allows you to change the old code to first emit a deprecation and then call the new code

@nofaralfasi would you mind responding to this?

ekohl reviewed (changes requested):
I'd want to get rid of using an instance of the class. That's the whole concept I'm urging you to deprecate. Here's now I'd implement it:

```ruby
class OwnerClassifier
  def initialize(id_and_type)
    @id_and_type = id_and_type
  end

  def user_or_usergroup
    Foreman::Deprecation.deprecation_warning("3.12", "`user_or_usergroup` is deprecated, use `OwnerClassifier.classify_owner` instead.")

    self.classify_owner(@id_and_type)
  end

  def self.classify_owner(id_and_type)
    return nil if id_and_type.blank?

    unless id_and_type.is_a?(String) && id_and_type.match?(/^\d+-(Users|Usergroups)$/)
      raise ArgumentError, _("Invalid input format. Please use the format '${id}-[Users|Usergroups]'.")
    end

    cls = id_and_type.end_with?('Users') ? User : Usergroup
    cls.find_by!(id_and_type.to_i)
  end
end
```

First, it uses `find_by_id!` to let Rails raise the NotFound exception. This simplifies the code and probably gives a more accurate exception as well.

The bigger style difference is that it removes all the helper methods. This is to remove all state in the class so that eventually we can remove `initialize()` and the whole `.new` invocation. Otherwise you still maintain the instance every single time and you get all the overhead of a class, with the downsides of also having a static method.

The `cls` variable is just to make the code a bit easier to understand. The `end_with?` is an unimportant micro optimization.

nofaralfasi commented:
> I'd want to get rid of using an instance of the class. That's the whole concept I'm urging you to deprecate. 

While I was trying to keep it, due to other recommendations I received in the review :) 
Thank you for the detailed explanation!

ekohl commented on app/services/owner_classifier.rb:
```diff
@@ -3,8 +3,24 @@ def initialize(id_and_type)
+    validate_input_format!(id_and_type)
+
+    owner_type = id_and_type.end_with?('Users') ? User : Usergroup
+    owner_type.find(id_and_type.to_i)
```
Shouldn't this be `find_by_id` like it was? Or `find_by_id!` to raise an exception?

ekohl replied:
Ok, I looked it up myself. https://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find states it will call `.to_i` for you so it can even be:
```suggestion
    owner_type.find(id_and_type)
```

But your code is also valid. I don't mind it being explicit here.

nofaralfasi replied:
AFAIK, both `find_by_id!` and `find` raise the same exception when the record is not found (in this case). If you believe that `find_by_id!` enhances readability in this context, I'm open to using it instead.

ekohl commented on test/unit/owner_classifier_test.rb:
```diff
@@ -2,24 +2,24 @@
 
 class OwnerClassifierTest < ActiveSupport::TestCase
```
Can you keep a test for the deprecated way? That it at least still works.
